src/zeroD/ReactorNet.cpp

```diff
@@ -242,51 +242,74 @@ double ReactorNet::advance(double time, bool applylimit)
+    m_ybase_time = m_time;
+    m_limit_check_active = true;
+    m_integ->setRootFunctionCount(nRootFunctions());
 
-    getAdvanceLimits(m_advancelimits.data());
-
-    // ensure that gradient is available
-    while (lastOrder() < 1) {
-        step();
+    // Integrate toward the requested time; integrator will return early if a limit is
+    // reached (CV_ROOT_RETURN)
+    try {
+        m_integ->integrate(time);
+    } catch (...) {
+        m_limit_check_active = false;
+        m_integ->setRootFunctionCount(nRootFunctions());
+        throw;
     }
+    m_time = m_integ->currentTime();
+
+    // Update reactor states to match the integrator solution at the time reached
+    // (which may be earlier than 'time' if a limit was triggered)
+    updateState(m_integ->solution());
 
-    int k = lastOrder();
-    double t = time, delta;
-    double* y = m_integ->solution();
+    // Disable limit checking after this call
+    m_limit_check_active = false;
+    m_integ->setRootFunctionCount(nRootFunctions());
 
-    // reduce time step if limits are exceeded
-    while (true) {
-        bool exceeded = false;
-        getEstimate(t, k, &m_yest[0]);
+    // When a root event stopped integration before reaching the requested time, report
+    // the most limiting component and details about the step.
+    if (m_verbose && m_time < time) {
+        // Ensure limits are available
+        if (m_advancelimits.size() != m_nv) {
+            m_advancelimits.assign(m_nv, -1.0);
+        }
+        getAdvanceLimits(m_advancelimits.data());
+        double* ycurr = m_integ->solution();
+        size_t jmax = npos;
+        double max_ratio = -1.0;
+        double best_delta = 0.0;
+        double best_limit = 0.0;
         for (size_t j = 0; j < m_nv; j++) {
-            delta = abs(m_yest[j] - y[j]);
-            if ( (m_advancelimits[j] > 0.) && ( delta > m_advancelimits[j]) ) {
-                exceeded = true;
-                if (m_verbose) {
-                    writelog("    Limiting global state vector component {:d} (dt = {:9.4g}):"
-                             "{:11.6g} > {:9.4g}\n",
-                             j, t - m_time, delta, m_advancelimits[j]);
+            double lim = m_advancelimits[j];
+            if (lim > 0.0) {
+                double delta = abs(ycurr[j] - m_ybase[j]);
+                double ratio = delta / lim;
+                if (ratio > max_ratio) {
+                    max_ratio = ratio;
+                    jmax = j;
+                    best_delta = delta;
+                    best_limit = lim;
                 }
             }
         }
-        if (!exceeded) {
-            break;
+        if (jmax != npos) {
+            double dt = m_time - m_ybase_time;
+            writelog("    Advance limit triggered for component {:d} (dt = {:9.4g}):"
+                     " |Δ| = {:11.6g}, limit = {:9.4g}, ratio = {:9.4g}\n",
+                     jmax, dt, best_delta, best_limit, max_ratio);
         }
```

Comment:
As long as CVODES is working correctly, some of the values printed here are not very interesting. From the example of `reactor1.py`, `best_delta` will always equal `best_limit`, and `max_ratio` should always be 1.0.